components/forwarder/forwarder.go

```
@@ -0,0 +1,88 @@
+	config.setDefaults()
+
+	router, err := message.NewRouter(message.RouterConfig{}, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &Forwarder{router, publisherOut, logger, config}
+
+	router.AddNoPublisherHandler(
+		"events_forwarder",
+		config.ForwarderTopic,
+		subscriberIn,
+		f.forwardMessage,
+	)
+
+	return f, nil
+}
+
+// Run runs forwarder's handler responsible for forwarding messages.
+// This call is blocking while the forwarder is running.
+// ctx will be propagated to the forwarder's subscription.
+//
+// To stop Run() you should call Close() on the forwarder.
+func (f *Forwarder) Run(ctx context.Context) error {
+	return f.router.Run(ctx)
+}
+
+// Close stops forwarder's handler.
+func (f *Forwarder) Close() error {
+	return f.router.Close()
+}
+
+// Running returns channel which is closed when the forwarder is running.
+func (f *Forwarder) Running() chan struct{} {
+	return f.router.Running()
+}
+
+// DecoratePublisher is a helper method which decorates a given publisher so when it publishes messages,
+// it envelopes them and publish directly to the forwarder's topic provided in the forwarder's config.
+//
+// Enveloping means that the message is put into the generic envelope containing the original message
+// and a destination topic taken from publisher `Publish` method. The destination topic is used later
+// by the forwarder to forward it to a specific topic.
+func (f *Forwarder) DecoratePublisher(publisher message.Publisher) message.Publisher {
+	return NewPublisherDecorator(publisher, f.config)
+}
+
+// AddSubscriberDecorators wraps the forwarder's Subscriber.
+func (f *Forwarder) AddSubscriberDecorators(dec ...message.SubscriberDecorator) {
+	f.router.AddSubscriberDecorators(dec...)
+}
+
+func (f *Forwarder) forwardMessage(msg *message.Message) error {
+	destTopic, unwrappedMsg, err := unwrapMessageFromEnvelope(msg)
```

what if someone uses the undecorated input publisher?
I guess we would get an error here.
Will the faulty message NACK and retry in eternity?

I guess that's up to the middlewares of the router, see my other comment.

Alternatively, we could detect undecorated messages and silently ignore them, but maybe that's too opinionated.

Either way, sounds like a test case to me :)

---

I have added a note regarding that and also covered this with a test case: 
a361c10